src/Datadog.Trace/Agent/AgentWriter.cs

```diff
@@ -122,13 +122,21 @@ public async Task FlushAndCloseAsync()
 
             var delay = Task.Delay(TimeSpan.FromSeconds(20));
 
-            var completedTask = await Task.WhenAny(_serializationTask, delay)
-                .ConfigureAwait(false);
+            // Internally WhenAny will make a defensive copy of the array
+            // so we reuse the outer one.
+            // https://source.dot.net/#System.Private.CoreLib/Task.cs,6028
+            // https://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,6478
+            Task[] whenAnyArray = new Task[2];
```

Comment:
Note that this is counter-productive on .NET 5, where they added a `Task.WhenAny` overload that takes two parameters: https://source.dot.net/#System.Private.CoreLib/Task.cs,6054

But we don't target .NET 5 yet, so good for now.

Reply:
See the conversation/change in https://github.com/DataDog/dd-trace-dotnet/pull/1306#discussion_r604472591 😄

Reply:
Yeah, we already had this conversation :D I'll add the `#if`.

Reply:
done in 40744c8

Reply:
Careful, you switched to WhenAll at one point there 😉

Reply:
Right, sorry 😄  Fixed.